pkg/expr/expr.go

```diff
@@ -0,0 +1,365 @@
+}
+
+func process(ex ast.Expr, tgt *Expression) (err error) {
+	switch v := ex.(type) {
+	case *ast.UnaryExpr:
+		tgt.Fn = &Function{Name: tMap[v.Op]}
+		if err = processFunc(tgt.Fn, []ast.Expr{v.X}); err != nil {
+			return
+		}
+	case *ast.BinaryExpr:
+		tgt.Fn = &Function{Name: tMap[v.Op]}
+		if err = processFunc(tgt.Fn, []ast.Expr{v.X, v.Y}); err != nil {
+			return
+		}
+	case *ast.CallExpr:
+		vfunc, found := v.Fun.(*ast.Ident)
+		if !found {
+			return fmt.Errorf("unexpected expression: %#v", v.Fun)
+		}
+		tgt.Fn = &Function{Name: vfunc.Name}
+		if err = processFunc(tgt.Fn, v.Args); err != nil {
+			return
+		}
+	case *ast.ParenExpr:
+		if err = process(v.X, tgt); err != nil {
+			return
+		}
+	case *ast.BasicLit:
+		tgt.Const, err = newConstant(v.Value, typeMap[v.Kind])
+		if err != nil {
+			return
+		}
+	case *ast.Ident:
+		// true and false are treated as identifiers by parser
+		// we need to convert them into constants here
+		lv := strings.ToLower(v.Name)
+		if lv == "true" || lv == "false" {
+			typedVal := true
+			if lv == "false" {
+				typedVal = false
+			}
+			tgt.Const = &Constant{Value: lv, Kind: config.BOOL, TypedValue: typedVal}
+		} else {
+			tgt.Var = &Variable{Name: v.Name}
+		}
+	case *ast.SelectorExpr:
+		// a.b.c.d is a selector expression
+		// normally one walks down a chain of objects
+		// we have chosen an internally flat namespace, therefore
+		// a.b.c.d if an identifer. converts
+		// a.b.c.d --> $a.b.c.d
+		// for selectorExpr length is guaranteed to be at least 2.
+		var w []string
+		if err = processSelectorExpr(v, &w); err != nil {
+			return
+		}
+		var ww bytes.Buffer
+		ww.WriteString(w[len(w)-1])
+		for idx := len(w) - 2; idx >= 0; idx-- {
+			ww.WriteString("." + w[idx])
+		}
+		tgt.Var = &Variable{Name: ww.String()}
+	case *ast.IndexExpr:
+		// accessing a map
+		// request.header["abc"]
+		tgt.Fn = &Function{Name: tMap[token.LBRACK]}
+		if err = processFunc(tgt.Fn, []ast.Expr{v.X, v.Index}); err != nil {
+			return
+		}
+	default:
+		return fmt.Errorf("unexpected expression: %#v", v)
+	}
+
+	return nil
+}
+
+func processSelectorExpr(exin *ast.SelectorExpr, w *[]string) (err error) {
+	ex := exin
+	for {
+		*w = append(*w, ex.Sel.Name)
+		switch v := ex.X.(type) {
+		case *ast.SelectorExpr:
+			ex = v
+		case *ast.Ident:
+			*w = append(*w, v.Name)
+			return nil
+		default:
+			return fmt.Errorf("unexpected expression: %#v", v)
+		}
+	}
+}
+
+func processFunc(fn *Function, args []ast.Expr) (err error) {
+	fAargs := []*Expression{}
+	for _, ee := range args {
+		aex := &Expression{}
+		fAargs = append(fAargs, aex)
+		if err = process(ee, aex); err != nil {
+			return
+		}
+	}
+	fn.Args = fAargs
+	return nil
+}
+
+// Parse parses a given expression to ast.Expression.
+func Parse(src string) (ex *Expression, err error) {
+	a, err := parser.ParseExpr(src)
+	if err != nil {
+		return nil, fmt.Errorf("parse error: %s %s", src, err)
+	}
+	glog.V(2).Infof("\n\n%s : %#v\n", src, a)
```

Comment:
`"\n\n%s : %#v\n"` => `parsed '%s' into '%#v'`

glog handles newlines for us and the line as written doesn't provide any context